src/django_idom/views.py

```diff
@@ -0,0 +1,7 @@
+from django.http import HttpRequest, HttpResponse
+from idom.config import IDOM_WED_MODULES_DIR
+
+
+def web_modules_file(request: HttpRequest, file: str) -> HttpResponse:
+    file_path = IDOM_WED_MODULES_DIR.current.joinpath(*file.split("/"))
+    return HttpResponse(file_path.read_text(), content_type="text/javascript")
```

Comment:
Rendering javascript dynamically is frowned upon in Django for performance reasons. Plus we're hitting the filesystem on every function call, which is horrible for IOPS.

Can I get a TLDR why we need these web modules to be dynamic?

Reply:
There might be a way to do this since `IDOM_WED_MODULES_DIR` can be changed. However files get added to the `IDOM_WED_MODULES_DIR` dynamically at runtime, so we'll need to know the directory where the static JS files will ultimately be served from.

Reply:
It actually looks like this might be another thing we could manage for the user by automatically placing the dynamically added files within `STATIC_ROOT`. I might not be understanding how [`collectstatic`](https://docs.djangoproject.com/en/3.2/ref/contrib/staticfiles/#django-admin-collectstatic) works though.

Reply:
In order for `collectstatic` to work, the files need to be present during Django's boot up sequence.

I still don't understand what web modules are, but can we compile them during build time? If so, then Django's staticfile system would be able to serve them.

Reply:
For context, "web module" is't really a technical term, it's just what I've been calling Javascript files that expose an interface IDOM expects. The `IDOM_WEB_MODULES_DIR` then, is the directory where these bundles get collected at runtime. Thus, the full set of files within `IDOM_WEB_MODULES_DIR` are not known at build-time.

With that said, the location of the directory can be changed. So if we know where within `STATIC_ROOT` we should drop these JS files such that they won't clash with those compiled from `collectstatic`, then I think we can make this work. In a sense, IDOM has a similar concept of static file collection, but one which happens later than Django's.

Reply:
You absolutely don't want to ever manually place files into `STATIC_ROOT`. That would break schemas such as the collectstatic [manifest backend](https://docs.djangoproject.com/en/3.2/ref/contrib/staticfiles/#django.contrib.staticfiles.storage.ManifestStaticFilesStorage). Additionally, it would be completely incompatible with python static file servers such as [whitenoise](https://github.com/evansd/whitenoise).

Does this **need** to be done at runtime? Is it possible to somehow have a command within IDOM to pre-compile these components at build time?

Reply:
There are several ways to load JS components in IDOM:

1. [dynamically](https://idom-docs.herokuapp.com/docs/javascript-components.html#dynamically-loaded-components)
2. [statically via PyPI](https://idom-docs.herokuapp.com/docs/javascript-components.html#distributing-javascript-via-pypi)
3. [from some url](https://idom-docs.herokuapp.com/docs/javascript-components.html#distributing-javascript-via-cdn) (usually a CDN)

Static file collection isn't possibly for (1), is possible for (2) but would require breaking changes in IDOM, and isn't needed for (3).

I'm not sure there's a great work around for this at the moment besides recommending that people who need performance use method (3) and serve their JS from a CDN.

Reply:
### My Thoughts
- **Option 1**
   - I'm assuming this the currently developed solution within `django-idom`.
   - Not ideal for this repo, causes large performance impact and increases IOPS unnecessarily.
   - If we set the `IDOM_WED_MODULES_DIR` to `django-idom/static/js`, is there any way we can get it to build early?
      - Does the IDOM repo have a function handler for creating web modules?
      - If so then we could potentially call this in `django_idom`'s `__init__` which is guaranteed to occur before collecstatic happens.
- **Option 2 (Preferred)**
   - This is ideal from a design perspective, but if there isn't any easy workaround for this then I recommend we just open an issue ticket for `django-idom` and handle it later.
- **Option 3**
   - Some Django users really like self hosting all their own files, but prepackaging JS on a CDN isn't the worst option. Would require us to reintroduce the `{% idom_scripts %} template tag.

Reply:
I think we should organize a quick chat to better explain all the working pieces here.

We might be able to use the templates and the `idom_component` declarations to find all the modules where web modules are defined. There'd still be some edge cases potentially, but pretty unlikely.

Reply:
I'm available today, and after 5 tomorrow.

Reply:
I won't be available this weekend or the next. Monday would work though.

Reply:
I'll be available Monday after 5, does that work?

Reply:
Will make a calendar event when I find time.

Reply:
This conversation is now being tracked here: https://github.com/idom-team/django-idom/issues/4